File: Driver.java
/****************************************************************************

Matthew Roth, A15685519
CSE 12, Winter 2019
February 19, 2019
cs12xcm
Assignment 7

File Name:      Driver.java

Description:    The driver will implement and demonstrate the Tree 
	program. The Driver has an inner class UCSDStudent to test the
	Tree program. 
***************************************************************************/

import java.io.*;

/**
 * This class will test the Tree program. The main
 * data fields are the name and the student number. 
 */
class UCSDStudent extends Base {
		
		//the name of the student and number for student 
        private String name;
        private long studentnum;
		
		/**
		 * the constructor to take name and student
		 * number as parameters
		 */
        public UCSDStudent (String nm, long stunum) {
                name = new String (nm);
                studentnum = stunum;
        }
		
		/**
		 * 
		 *	return the name of the student
		 */
        public String getName () {
                return name;
        }
		
		/**
		 * the equals method to compare objects 
		 * of the UCSDStudent type 
		 *
		 */
        public boolean equals (Object object) {
		
		//check if the object is same type as this
		if (this == object)
			return true;
		
		//check if instanceof UCSDStudent  		
		if (!(object instanceof UCSDStudent))
			return false;
		
		//cast to a UCSDStudent 
		UCSDStudent otherVar = (UCSDStudent) object;
		
		return name.equals (otherVar.getName ());

        }
		
		/**
		 * this method will check which name is 
		 * higher precedence
		 *
		 */
        public boolean isGreaterThan (Base base) {
                return (name.compareTo (base.getName ()) > 0) ? true : false;
        }

		/**
		 * the toString method to return a string 
		 * representation of the object
		 *
		 */
        public String toString () {
                return "name:  " + name + "  studentnum:  " + studentnum;
        }
}

//PROVIDED CODE BELOW
public class Driver {
        private static final short NULL = 0;

        public static void main (String [] args) {
        
                /* initialize debug states */
                Tree.debugOff();

                /* check command line options */
                for (int index = 0; index < args.length; ++index) {
                        if (args[index].equals("-x"))
                                Tree.debugOn();
                }


                /* The real start of the code */
                SymTab<UCSDStudent> symtab = 
                                new SymTab<UCSDStudent>("UCSDStudentTree");
                String buffer = null;
                char command;
                long number = 0;

                System.out.println ("Initial Symbol Table:\n" + symtab);

                while (true) {
                        command = NULL; // reset command each time in loop
                        System.out.print ("Please enter a command:  " + 
                                "((a)llocate, is(e)mpty, (i)nsert, (l)ookup,"+
                                " (r)emove, (w)rite):  ");

                        try {
                        command = MyLib.getchar ();
                        MyLib.clrbuf (command); // get rid of return

                        switch (command) {
                        case 'a':
                                System.out.print
                                ("Please enter name of new Tree to " +
                                 "allocate:  ");
                                
                                buffer = MyLib.getline ();// formatted input
                                symtab = new SymTab<UCSDStudent>(buffer);
                                break;
			case 'e':
				if(symtab.isEmpty()){
					System.out.println("Tree is empty.");
				} else {
					System.out.println("Tree is "+
						"not empty.");
				}
				break;

                        case 'i':
                                System.out.print
                                ("Please enter UCSD student name to insert:  ");

                                buffer = MyLib.getline ();// formatted input

                                System.out.print
                                        ("Please enter UCSD student number:  ");

                                number = MyLib.decin ();
                                MyLib.clrbuf (command); // get rid of return

                                // create student and place in symbol table
                                symtab.insert(new UCSDStudent (buffer, number));
                                break;

                        case 'l': { 
                                UCSDStudent found;      // whether found or not

                                System.out.print
                                ("Please enter UCSD student name to lookup:  ");
                                buffer = MyLib.getline ();// formatted input

                                UCSDStudent stu = new UCSDStudent (buffer, 0);
                                found = symtab.lookup (stu);
                                
                                if (found != null) {
                                        System.out.println("Student found!");
                                        System.out.println(found);
                                }
                                else
                                        System.out.println ("student " + buffer
                                                + " not there!");
                                }
                                break;
                        
                        case 'r': { 
                                UCSDStudent removed; // data to be removed

                                System.out.print
                                ("Please enter UCSD student name to remove:  ");

                                buffer = MyLib.getline ();

                                UCSDStudent stu = new UCSDStudent (buffer, 0);

                                removed = symtab.remove(stu);

                                if (removed != null) {
                                        System.out.println("Student removed!"); 
                                        System.out.println(removed);
                                }
                                else
                                        System.out.println ("student " + buffer
                                                + " not there!");
                                }
                                break;

                        case 'w':
                                System.out.println("The Symbol Table " +
                                "contains:\n" + symtab);
                        }
                        }
                        catch (EOFException eof) {
                                break;
                        }
                }
                System.out.println("\nFinal Symbol Table:\n" + symtab);
        }
}

File: Driver.c
//==========================================================================
// cs12xcm                         Homework 7            Matt Roth 
//
// February 19 2019	
//--------------------------------------------------------------------------
// File: Driver.c
//
// Description:		The driver will implement and demonstrate the Tree 
//		program. The Driver has an inner class UCSDStudent to test the
//		Tree program. 
//==========================================================================

#include <cstdlib>
#include <iostream>
#include <string.h>
#include "Base.h"
#include "SymTab.h"
using namespace std;

#ifdef NULL
#undef NULL
#define NULL 0
#endif

//=========================================================================
// class UCSDStudent 
//
// Description:		This class will test the Tree program. The main
//				data fields are the name and the student number. 
//
//  
// Data Fields:
//
//     name - the name of the student    
//     studentnum - the number of the student 
//     
//
// Public functions:
//					UCSDStudent - constructor(s) for initialization
//					~UCSDStudent - destructor 
//					ostream & Write - the toString  
//					operator const char * - overrride the char *   
//					operator  - override the == 
//					operator > - override the >  
//==========================================================================
class UCSDStudent : public Base {

	//the name and the student number 
	char * name;
	long studentnum;

public:
	
/***************************************************************************
% Routine Name : UCSDStudent (public)
% File :         Driver.c
% 
% Description :  initialize the data fields
***************************************************************************/

    UCSDStudent(char * nam, long stunum = 0):
		name(strdup (nam)), studentnum(stunum){}

/***************************************************************************
% Routine Name : UCSDStudent (public)
% File :         Driver.c
% 
% Description :  initialize the data fields
***************************************************************************/
	UCSDStudent (const UCSDStudent & ucsdstudent){
		name = strdup (ucsdstudent.name);
		studentnum = ucsdstudent.studentnum;
	}

/***************************************************************************
% Routine Name : ~UCSDStudent (public)
% File :         Driver.c
% 
% Description :  Deallocated memory deletes the data 
***************************************************************************/
	~UCSDStudent (void){
		free(name);
	}

/***************************************************************************
% Routine Name : const char * (public)
% File :         Driver.c
% 
% Description :  overload the char * 
***************************************************************************/
	operator const char * (void) const{
		return name;
	}

/***************************************************************************
% Routine Name : operator == (public)
% File :         Driver.c
% 
% Description :  overload the == 
***************************************************************************/
	long operator == (const Base & bbb) const {
		return ! strcmp (name, bbb);
	}

/***************************************************************************
% Routine Name : operator > (public)
% File :         Driver.c
% 
% Description :  overload the > 
***************************************************************************/
	long operator > (const Base & bbb) const {
		return (strcmp (name, bbb) > 0) ? 1 : 0;
	}

/***************************************************************************
% Routine Name : ostream & Write (public)
% File :         Driver.c
% 
% Description :  toString 
***************************************************************************/
	ostream & Write (ostream & stream) const {
		return stream << "name:  " << name
			<< "  studentnum:  " << studentnum;
	}
};


//PROVIDED CODE BELOW
int main (int argc, char * const * argv) {
	char buffer[80];
	char command;
	long number;

	Tree::Set_Debug(0);

	if (argc != 1 && strcmp("-x", argv[1]) == 0) {
		Tree::Set_Debug(1);
	}
	
	SymTab * ST;
	ST = new SymTab ("UCSDStudentTree");
	ST->Write (cout << "Initial Symbol Table:\n");

	while (cin) {
		command = NULL;		// reset command each time in loop
		cout <<  "Please enter a command:  ((a)llocate, is(e)mpty," <<
			" (i)nsert, (l)ookup, (r)emove, (w)rite):  ";
		cin >> command;

		switch (command) {

		case 'a':
			cout << "Please enter name of new Tree to allocate:  ";
			cin >> buffer; 

			delete ST;
			ST = new SymTab (buffer);

			break;

		case 'e':
                        if (ST->IsEmpty ())
                                cout << "Tree is empty." << endl;
                        else
                                cout << "Tree is not empty." << endl;
                        break;
			
		case 'i':
			cout << "Please enter UCSD student name to insert:  ";
			cin >> buffer;	// formatted input

			cout << "Please enter UCSD student number:  ";
			cin >> number;

			// create student and place in symbol table
			ST->Insert (new UCSDStudent (buffer, number));
			break;

		case 'l': {
			const Base * found;	// whether found or not

			cout << "Please enter UCSD student name to lookup:  ";
			cin >> buffer;	// formatted input

			UCSDStudent stu (buffer, 0);
			found = ST->Lookup (&stu);
			
			if (found)
				found->Write (cout << "Student found!\n") << "\n";
			else
				cout << "student " << buffer << " not there!\n";
			}
			break;
		
		case 'r': {
			Base * removed;	// data to be removed

			cout << "Please enter UCSD student name to remove:  ";
			cin >> buffer;	// formatted input

			UCSDStudent stu (buffer, 0);
			removed = ST->Remove (&stu);

			if (removed)
				removed->Write (cout << "Student removed!\n") << "\n";
			else
				cout << "student " << buffer << " not there!\n";
			}
			break;


		case 'w':
			ST->Write (cout << "The Symbol Table contains:\n");
		}
	}

	ST->Write (cout << "\nFinal Symbol Table:\n");
	delete ST;
}
